readme.md
# l2met

Convert a formatted log stream into metrics.

* [Current Release](#current-release)
* [Synopsis](#synopsis)
* [Features](#features)
* [Log Conventions](#log-conventions)
* [Getting Started](#getting-started)
* [Operations](#operations)
* [Hacking on l2met](#hacking-on-l2met)

## Current Release

* Tip: [v2.0beta](https://github.com/ryandotsmith/l2met/tree/v2.0beta)
* Stable: [v1.6](https://github.com/heroku/l2met/tree/v1.6)

## Synopsis

L2met receives HTTP requests that contain a body of rfc5424 formatted data. Commonly data is drained into l2met by [logplex](https://github.com/heroku/logplex) or [log-shuttle](https://github.com/ryandotsmith/log-shuttle).

Once data is delivered, l2met extracts and parses the individual log lines using the [log conventions](#log-conventions) and then stores the data in redis so that outlets can read the data and build metrics. The librato_outlet is the most popular and will put all of your metrics into your Librato account. See the [getting started](#getting-started) section to get started.

## Log Conventions

L2met uses convention over configuration to build metrics. Keys that are prefixed with `measure.` and have numerical values will be analyzed. L2met uses a loosely defined key=val log structure.

### Keywords:

There are only a few keywords that l2met will parse.

* measure.*
* source

Any key that begins with `measure.` will be considered for a measurement. You can add a source key to further identify the measurement.

A simple example:

```ruby
$stdout.puts("measure.db.latency=20")
```

Metrics Produced:

* `db.latency.{min, median, perc95, perc99, max, mean, last, count, sum}`

An example using [multi-metrics](#multi-metrics) and a source key:

```ruby
$stdout.puts("source=prod measure.db.latency=20 measure.view.latency=10")
```

Metrics Produced:

* `prod.db.latency.{min, median, perc95, perc99, max, mean, last, count, sum}`
* `prod.view.latency.{min, median, perc95, perc99, max, mean, last, count, sum}`

## Features

* [High resolution buckets](#high-resolution-buckets)
* [Drain Prefix](#drain-prefix)
* [Multi-metrics](#multi-metrics)
* [Heroku Router](#heroku-router)
* [Bucket attrs](#bucket-attrs)
* [HTTP Outlet](#http-outlet)
* Librato Outlet

### Multi-metrics

We want to be able to specify multiple measurements on a single line so as not to have to pay the (albeit low) overhead of writing to stdout. However, we don't want to take every k=v under the sun. L2met has always forced you to think about the things you are measuring and this feature does not regress in that regard.

Example:

```
echo 'measure.hello=10 measure.world=10' | log-shuttle
```

This will result in 2 buckets:

1. {name=hello, vals=[10], ...}
2. {name=world, vals=[10], ...}

Thus you can measure multiple things provided the key is prefixed with `measure.`.

### Heroku Router

The Heroku router has a log line convention described [here](https://devcenter.heroku.com/articles/http-routing#heroku-router-log-format).

This feature will read the User field in the syslog packet looking for the string "router." If a match is had, we will parse the structured data and massage it to match the l2met convention. Furthermore, we will prefix the measurement name with the parsed host field in the log line.

Example:

```
path=/logs host=test.l2met.net connect=1ms service=2ms bytes=0
```

Would produce the following buckets:

1. {name=router.connect, source=test.l2met.net.connect, vals=[1]}
2. {name=router.service, source=test.l2met.net.service, vals=[2]}
3. {name=router.bytes,   source=test.l2met.net.bytes,   vals=[0]}

### Drain Parameters

There are several configuration options that can be specified at the drain level. These options will be applied to all of the metrics coming into the drain.

* Resolution
* Prefix

#### High Resolution Buckets

By default, l2met will hold measurements in 1 minute buckets. This means that data visible in Librato and Graphite have a granularity of 1 minute. However, it is possible to achieve a greater level of resolution. For example, you can get 1 second level resolution by appending a query parameter to your drain url. Note that the resolution is specified in seconds.

The supported resolutions are: 1, 5, 30, 60.

Drain URL:

```
https://user:token@l2met.herokuapp.com/logs?resolution=1
```

#### Drain Prefix

It can be useful to prepend a string to all metrics going into a drain. For instance, say you want all of your metrics to include the environment or app name. You can achieve this by adding a drain prefix.

Drain URL:

```
https://token@l2met.herokuapp.com/logs?prefix=myapp.staging
```

### Bucket Attrs

Most outlet providers allow meta-data to be associated with measurements. Bucket attrs is the l2met feature that allows you to interact with this meta-data.

#### Units

L2met supports associating units with numbers by appending a non-digit sequence after the digits. L2met will assign all measurements a units of *y* unless specified. You can specify a unit by prepending `/[a-zA-z]+/` after the digit. For instance:

```
measure.db.get=1ms measure.web.get=1
```

This will create the following buckets:

1. {name=db.get,  vals=[1], units=ms}
2. {name=web.get, vals=[1], units=y}

#### Min Y Value

Librato allows charts to have a min y value. L2met sets this to 0. It cannot be overridden at this time. Open a GH issue if this is a problem for you.

### HTTP Outlet

The HTTP Outlet is a read API for your metrics. You can query metrics by id. Id construction is a bit rough at this stage, but if you know what you are looking for, reading metrics can be quite simple. For example, if the following logs are emitted:

```
measure.db.get=10ms
```

The metrics can be accessed by the following request:

```bash
$ curl http://your-token@l2met.net/metrics?name=db.get&resolution=60&units=ms&limit=1&offset=1
```

Depending on the resolution of the drain, you will get the last bucket offset by one. The offset implies that you are reading the previous bucket with respect to time. For instance, if your resolution is 60 (1 minute) then an offset of 1 will produce the last minute's bucket.

#### Metric Assertion

You can also assert what you expect the metrics to be. This is useful in that you can point pingdom directly at your l2met instance and create alerts on metrics. For example, if an alert must be made when user signups fall below 5 per minute, we could construct the following request:

```bash
$ curl http://your-token@l2met.net/metrics?name=user.signup&resolution=60&limit=1&offset=1&count=5
```

If there were only 4 counts of user.signup, then the query would return a 404 which would cause pingdom to send the alert.

The following metrics can be asserted:

* count
* mean
* sum

Tolerance can also be applied to all of the assertions. This compensates for the lack of inequality operators in the URL. If the tolerance is not present, it is assumed to be 0. If the tolerance is specified, it applies to all metric assertions. You must construct separate requests if you want unique tolerance to metric assertions.

For example, if we wanted to test if the count was 5 +/- 2, the following request would suffice:

```bash
$ curl http://your-token@l2met.net/metrics?name=user.signup&resolution=60&limit=1&offset=1&count=5&tol=2
```

## Getting Started

The easiest way to get l2met up and running is to deploy to Heroku. This guide assumes you have already created a Heroku & Librato account.

```bash
$ curl https://s3-us-west-2.amazonaws.com/l2met/v2.0beta/linux/amd64/l2met.tar.gz | tar xvz

$ ./setup my-l2met e@foo.com abc123
...
Drain URL: https://long-token@my-l2met.herokuapp.com/logs
```

This command will create Heroku app named `my-l2met` and return a drain URL with encrypted Librato credentials for a Librato account with email `e@foo.com` and an API token of `abc123`. After you have created `my-l2met`, you can add the drain URL to a Heroku app. A copy of the log stream will be delivered to `my-l2met` and metrics will be sent to the Librato account which your provided in the setup.

```bash
$ heroku drains:add https://long-token@my-l2met.herokuapp.com/logs -a myapp
```

You can manually send data to `my-l2met` by the following curl command:

```bash
$ curl "https://long-token@my-l2met.herokuapp.com/logs" --data "94 <190>1 2013-03-27T20:02:24+00:00 hostname token shuttle - - measure.hello=99 measure.world=100"
```

Verify the command worked by viewing the [newly created metrics](https://metrics.librato.com/metrics?search=hello).

## Operations
There are several knobs and levers to use in order to optimize l2met operations.

* [Internal Metrics](#internal-metrics)

### Internal Metrics
This data is sent through the normal flow for metrics to the librato outlet. To enable, ensure the following config var is set:

```bash
$ export OUTLET_USER="user@domain.com" OUTLET_PASS="abc123"
```

The following internal metrics are produced:
* receiver latency & count: `name=l2met.receiver.receive`
* receiver drops: `name=l2met.receiver.drop`

## Hacking on l2met

Before working on a new feature, send your proposal to the [mailing list](https://groups.google.com/d/forum/l2met) for tips & feedback. Be sure to work on a feature branch and submit a PR when ready.

[![Build Status](https://drone.io/github.com/ryandotsmith/l2met/status.png)](https://drone.io/github.com/ryandotsmith/l2met/latest)

```bash
$ go version
go version go1.1rc2 darwin/amd64
```

```bash
$ git clone git://github.com/ryandotsmith/l2met.git
$ cd l2met
$ export OUTLET_USER="na" OUTLET_PASS="na" 
$ export SECRETS=$(dd if=/dev/urandom bs=32 count=1 2>/dev/null | openssl base64)
$ export TZ=UTC
$ go test ./...
```


changelog.md
Each version of l2met corresponds to a git tag.

## 1.7

* Disconnect the locking redis connection on error
* Log bucket reader partition locking and redis store scanning

## 1.6

* Fix for auth path that could never complete
* include dod@heroku.com account for complex measurements

## 1.5

* Add .mean|sum|count|min|max for non librato-production@heroku.com cases

## 1.4

* Drop .mean|sum|count|min|max in favor of complex measurements

## 1.2

2013-05-06

* improve logging for internal l2met metrics. now uses latest l2met convention

## 1.1

2013-05-06

* fix bugs in base64 decoded authentication parsing

## 1.0

2013-04-24

* remove token based auth. (No longer depending on PostgreSQL)

## 0.11

2013-04-21

* bugfix: correct http outlet resolution parsing
* bugfix: correct librato outlet's use of User-Agent
* bugfix: handle pg errors when fetching tokens
* Improve readme
* Make APP_NAME optional
* Add a receiver metric that will print number of buckets processed
* Improve test coverage

## 0.10

2013-04-10

* [HTTP Outlet](https://github.com/ryandotsmith/l2met#http-outlet)

## 0.9

2013-04-10

* bugfix: heroku router log lines were not being parsed. this is fixed.
* bugfix: heroku router and drain prefix features were incompatible.

## 0.8

2013-04-10

* First pass at Dbless Authentication. This does not include credential signing.
* Drain prefix

## 0.7

2013-04-05

* Fix regression related to parsing floating point numbers.

## 0.6

2013-04-04

* [Bucket Attrs](https://github.com/ryandotsmith/l2met#bucket-attrs)

## 0.5

2013-04-02

* Updated the resolution format to be specified in minutes.
* Improved the reporting around outlet delays.

## 0.4

2013-04-02

* [Heroku Router](https://github.com/ryandotsmith/l2met#heroku-router)

## 0.3

2013-04-02

* [Multi-metrics](https://github.com/ryandotsmith/l2met#multi-metrics)

## 0.2

2013-04-01

* PR [#34](https://github.com/ryandotsmith/l2met/pull/34)
* [High resolution buckets](https://github.com/ryandotsmith/l2met#high-resolution-buckets)

## 0.1

2013-03-25

* Init


main.go
package main

import (
	"fmt"
	"io/ioutil"
	"l2met/outlet"
	"l2met/receiver"
	"l2met/store"
	"l2met/utils"
	"log"
	"net/http"
	"os"
	"runtime"
	"time"
)

func init() {
	runtime.GOMAXPROCS(runtime.NumCPU())
}

func main() {
	// The number of partitions that our backends support.
	numPartitions := utils.EnvUint64("NUM_OUTLET_PARTITIONS", 1)
	// The bucket.Store struct will initialize a redis pool for us.
	maxRedisConn := utils.EnvInt("OUTLET_C", 2) + 100
	// We use the store to Put buckets into redis.
	server, pass, err := utils.ParseRedisUrl()
	if err != nil {
		log.Fatal(err)
	}
	rs := store.NewRedisStore(server, pass, numPartitions, maxRedisConn)

	reqBuf := utils.EnvInt("REQUEST_BUFFER", 1000)
	recv := receiver.NewReceiver(reqBuf, reqBuf)
	recv.FlushInterval = time.Millisecond * 200
	recv.NumOutlets = utils.EnvInt("OUTLET_C", 100)
	recv.NumAcceptors = utils.EnvInt("ACCEPT_C", 100)
	recv.Store = rs
	recv.Start()
	go recv.Report()

	http.HandleFunc("/health", func(w http.ResponseWriter, r *http.Request) {
		healthCheck(w, r, rs)
	})
	http.HandleFunc("/logs", func(w http.ResponseWriter, r *http.Request) {
		recvLogs(w, r, recv)
	})

	httpOutlet := new(outlet.HttpOutlet)
	httpOutlet.Store = rs
	http.HandleFunc("/metrics", func(w http.ResponseWriter, r *http.Request) {
		httpOutlet.ServeReadBucket(w, r)
	})

	port := utils.EnvString("PORT", "8000")
	err = http.ListenAndServe(":"+port, nil)
	if err != nil {
		fmt.Printf("error=%s msg=%q\n", err, "Unable to start http server.")
		os.Exit(1)
	}
	fmt.Printf("at=l2met-initialized port=%s\n", port)
}

func healthCheck(w http.ResponseWriter, r *http.Request, s store.Store) {
	ok := s.Health()
	if !ok {
		msg := "Redis is unavailable."
		fmt.Printf("error=%q\n", msg)
		http.Error(w, msg, 500)
	}
}

// Pull data from the http request, stick it in a channel and close the request.
// We don't do any validation on the data. Always respond with 200.
func recvLogs(w http.ResponseWriter, r *http.Request, recv *receiver.Receiver) {
	defer utils.MeasureT("http-receiver", time.Now())
	if r.Method != "POST" {
		http.Error(w, "Invalid Request", 400)
		return
	}
	user, pass, err := utils.ParseAuth(r)
	if err != nil || user == "" || pass == "" {
		fmt.Printf("measure.failed-auth error=%s user=%s pass=%s user-agent=%s token=%s client=%s\n",
			err, user, pass, r.Header.Get("User-Agent"), r.Header.Get("Logplex-Drain-Token"), r.Header.Get("X-Forwarded-For"))
		http.Error(w, "Invalid Request", 400)
		return
	}
	b, err := ioutil.ReadAll(r.Body)
	r.Body.Close()
	if err != nil {
		http.Error(w, "Invalid Request", 400)
		return
	}
	recv.Receive(user, pass, b, r.URL.Query())
}


utils/auth.go
package utils

import (
	"encoding/base64"
	"errors"
	"github.com/kr/fernet"
	"net/http"
	"os"
	"strings"
	"time"
)

var (
	OneHundredYears = time.Hour * 24 * 365 * 100
	keys            []*fernet.Key
)

func init() {
	if s := strings.Split(os.Getenv("SECRETS"), ":"); len(s) > 0 {
		keys = fernet.MustDecodeKeys(s...)
	}
}


func parseAuthHeader(r *http.Request) (string, error) {
	header, ok := r.Header["Authorization"]
	if !ok {
		return "", errors.New("Unable to parse Authorization header.")
	}
	return header[0], nil
}

func parseAuthValue(header string) (string, string ,error) {
	parts := strings.SplitN(header, " ", 2)
	if len(parts) != 2 {
		return "", "", errors.New("Authorization header malformed.")
	}

	method := parts[0]
	if method != "Basic" {
		return "", "", errors.New("Authorization must be basic.")
	}

	payload := parts[1]
	decodedPayload, err := base64.StdEncoding.DecodeString(payload)
	if err != nil {
		return "", "", err
	}

	userPass := strings.SplitN(string(decodedPayload), ":", 2)
	switch len(userPass) {
	case 1:
		return userPass[0], "", nil
	case 2:
		return userPass[0], userPass[1], nil
	}

	return "", "", errors.New("Unable to parse username or password.")
}

func ParseAuth(r *http.Request) (string, string, error) {
	header, err := parseAuthHeader(r)
	if err != nil {
		return "", "", err
	}

	user, _, err := parseAuthValue(header)
	if err != nil {
		return "", "", err
	}

	//If we have gotten here, we have a signed, db-less authentication reque
	//If we can verify and decrypt, then we will pass the decrypted credential
	//to the caller. Most of the time, the username and password will be
	//credentials to outlet providers. (e.g. Librato creds or Graphite creds
	//We care about the validity of those credentials here. If they are wrong
	//the metrics will be dropped at the outlet. Keep an eye on http
	//authentication errors from the log output of the outlets.
	if len(keys) > 0 {
		if s := fernet.VerifyAndDecrypt([]byte(user), OneHundredYears, keys); s != nil {
			parts := strings.Split(string(s), ":")
			return parts[0], parts[1], nil
		}
	}
	//If the user is not == "l2met" then dbless-auth is requested.
	//ATM we assume the first part (user field) contains a base64 encoded
	//representation of the outlet credentials.
	if len(user) > 0 {
		trimmedUser := strings.Replace(user, ":", "", -1)
		decodedUser, err := base64.StdEncoding.DecodeString(trimmedUser)
		if err != nil {
			return trimmedUser, "", err
		}
		outletCreds := strings.Split(string(decodedUser), ":")
		//If the : is absent in parts[0], outletCreds[0] will contain the entire string in parts[0].
		u := outletCreds[0]
		//It is not required for the outletCreds to contain a pass.
		//The empty string that is returned will be handled by the outlet.
		var p string
		if len(outletCreds) > 1 {
			p = outletCreds[1]
		}
		return u, p, nil
	}
	return "", "", errors.New("End of Authe chain reached.")
}


utils/lock.go
package utils

import (
	"errors"
	"fmt"
	"github.com/garyburd/redigo/redis"
	"log"
	"time"
)

var rc redis.Conn

func connect() {
	if rc != nil {
		return
	}

	fmt.Printf("ns=utils-lock fn=connect at=start\n")

	var err error
	host, password, err := ParseRedisUrl()
	if err != nil {
		log.Fatal(err)
	}
	rc, err = redis.Dial("tcp", host)
	if err != nil {
		log.Fatalf("Locking service is unable to connect to redis. err: %s",
			err)
	}
	rc.Do("AUTH", password)

	fmt.Printf("ns=utils-lock fn=connect at=finish\n")
}

func disconnect() {
	fmt.Printf("ns=utils-lock fn=disconnect at=start\n")
	rc.Close()
	rc = nil
	fmt.Printf("ns=utils-lock fn=disconnect at=finish\n")
}

func init() {
	connect()
}

func UnlockPartition(key string) {
	rc.Do("DEL", key)
}

func LockPartition(ns string, max, ttl uint64) (uint64, error) {
	for {
		for p := uint64(0); p < max; p++ {
			name := fmt.Sprintf("lock.%s.%d", ns, p)
			locked, err := writeLock(name, ttl)
			if err != nil {
				return 0, err
			}
			if locked {
				return p, nil
			}
		}
		time.Sleep(time.Second * 5)
	}
	return 0, errors.New("LockPartition impossible broke the loop.")
}

func writeLock(name string, ttl uint64) (bool, error) {
	connect()
	new := time.Now().Unix() + int64(ttl) + 1
	old, err := redis.Int(rc.Do("GETSET", name, new))
	// If the ErrNil is present, the old value is set to 0.
	if err != nil && err != redis.ErrNil && old == 0 {
		disconnect()
		return false, err
	}
	// If the new value is greater than the old
	// value, then the old lock is expired.
	return new > int64(old), nil
}


store/redis_store.go
package store

import (
	"fmt"
	"github.com/garyburd/redigo/redis"
	"hash/crc64"
	"l2met/bucket"
	"l2met/encoding"
	"l2met/utils"
	"time"
)

var PartitionTable = crc64.MakeTable(crc64.ISO)

type RedisStore struct {
	redisPool     *redis.Pool
	maxPartitions uint64
}

func NewRedisStore(server, pass string, maxPartitions uint64, maxConn int) *RedisStore {
	return &RedisStore{
		maxPartitions: maxPartitions,
		redisPool:     initRedisPool(server, pass, maxConn),
	}
}

func initRedisPool(server, pass string, maxConn int) *redis.Pool {
	return &redis.Pool{
		MaxIdle:     maxConn,
		IdleTimeout: 30 * time.Second,
		TestOnBorrow: func(c redis.Conn, t time.Time) error {
			_, err := c.Do("PING")
			return err
		},
		Dial: func() (redis.Conn, error) {
			c, err := redis.Dial("tcp", server)
			if err != nil {
				return nil, err
			}
			c.Do("AUTH", pass)
			return c, err
		},
	}
}

func (s *RedisStore) MaxPartitions() uint64 {
	return s.maxPartitions
}

func (s *RedisStore) Health() bool {
	rc := s.redisPool.Get()
	defer rc.Close()
	_, err := rc.Do("PING")
	if err != nil {
		return false
	}
	return true
}

func (s *RedisStore) Scan(partition string) <-chan *bucket.Bucket {
	retBuckets := make(chan *bucket.Bucket)
	go func(out chan *bucket.Bucket) {
		defer utils.MeasureT("redis-store.scan", time.Now())
		rc := s.redisPool.Get()
		defer rc.Close()
		defer close(out)
		rc.Send("MULTI")
		rc.Send("SMEMBERS", partition)
		rc.Send("DEL", partition)
		reply, err := redis.Values(rc.Do("EXEC"))
		if err != nil {
			fmt.Printf("at=%q error=%s\n", "bucket-store-scan", err)
			return
		}
		var delCount int64
		var members []string
		redis.Scan(reply, &members, &delCount)
		for _, member := range members {
			id, err := bucket.ParseId(member)
			if err != nil {
				fmt.Printf("at=%q error=%s\n", "bucket-store-parse-key", err)
				continue
			}
			out <- &bucket.Bucket{Id: id}
		}
	}(retBuckets)
	return retBuckets
}

func (s *RedisStore) Putback(partition string, id *bucket.Id) error {
	defer utils.MeasureT("bucket.putback", time.Now())
	rc := s.redisPool.Get()
	defer rc.Close()
	_, err := rc.Do("SADD", partition, id.String())
	return err
}

func (s *RedisStore) Put(b *bucket.Bucket) error {
	defer utils.MeasureT("bucket.put", time.Now())

	rc := s.redisPool.Get()
	defer rc.Close()

	b.Lock()
	key := b.Id.String()
	value := b.Vals
	b.Unlock()

	//TODO(ryandotsmith): Ensure consistent keys are being written.
	partition := s.bucketPartition("outlet", []byte(key))
	rc.Send("MULTI")
	rc.Send("RPUSH", key, value)
	rc.Send("EXPIRE", key, 300)
	rc.Send("SADD", partition, key)
	rc.Send("EXPIRE", partition, 300)
	_, err := rc.Do("EXEC")
	if err != nil {
		return err
	}
	return nil
}

func (s *RedisStore) Get(b *bucket.Bucket) error {
	defer utils.MeasureT("bucket.get", time.Now())

	rc := s.redisPool.Get()
	defer rc.Close()

	//Fill in the vals.
	reply, err := redis.Values(rc.Do("LRANGE", b.Id.String(), 0, -1))
	if err != nil {
		return err
	}
	for _, item := range reply {
		v, ok := item.([]byte)
		if !ok {
			continue
		}
		err = encoding.DecodeArray(v, &b.Vals, '[', ']', ' ')
	}
	return nil
}

func (s *RedisStore) bucketPartition(prefix string, b []byte) string {
	check := crc64.Checksum(b, PartitionTable)
	return fmt.Sprintf("%s.%d", prefix, check%s.MaxPartitions())
}


outlet/bucket_reader.go
package outlet

import (
	"fmt"
	"l2met/bucket"
	"l2met/store"
	"l2met/utils"
	"time"
)

type BucketReader struct {
	Store       store.Store
	Interval    time.Duration
	Partition   string
	Ttl         uint64
	NumOutlets  int
	NumScanners int
	Inbox       chan *bucket.Bucket
	Outbox      chan *bucket.Bucket
}

func NewBucketReader(mi int) *BucketReader {
	rdr := new(BucketReader)
	rdr.Partition = "bucket-reader"
	rdr.Inbox = make(chan *bucket.Bucket, mi)
	return rdr
}

func (r *BucketReader) Start(out chan *bucket.Bucket) {
	r.Outbox = out
	go r.scan()
	for i := 0; i < r.NumOutlets; i++ {
		go r.outlet()
	}
}

func (r *BucketReader) scan() {
	for _ = range time.Tick(r.Interval) {
		//TODO(ryandotsmith): It is a shame that we have to lock
		//for each interval. It would be great if we could get a lock
		//and work for like 1,000 intervals and then relock.
		p, err := utils.LockPartition(r.Partition, r.Store.MaxPartitions(), r.Ttl)
		if err != nil {
			fmt.Printf("ns=bucket-reader-scan step=lock-partition at=error error=%q\n", err)
			continue
		}
		fmt.Printf("ns=bucket-reader-scan step=lock-partition at=success partition=%d\n", p)
		partition := fmt.Sprintf("outlet.%d", p)
		for bucket := range r.Store.Scan(partition) {
			valid := time.Now().Add(bucket.Id.Resolution)
			//TODO(ryandotsmith): This seems ripe for a lua script.
			//The goal would to be receive data from scan that is sure
			//to be valid.
			if bucket.Id.Time.Before(valid) {
				r.Inbox <- bucket
			} else {
				if err := r.Store.Putback(partition, bucket.Id); err != nil {
					fmt.Printf("error=%s\n", err)
				}
			}
		}
		utils.UnlockPartition(fmt.Sprintf("%s.%d", r.Partition, p))
	}
}
func (r *BucketReader) outlet() {
	for b := range r.Inbox {
		r.Store.Get(b)
		r.Outbox <- b
	}
}


outlet/librato_outlet.go
package outlet

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"io/ioutil"
	"l2met/bucket"
	"l2met/utils"
	"net/http"
	"strconv"
	"time"
)

var libratoUrl = "https://metrics-api.librato.com/v1/metrics"

type LibratoAttributes struct {
	Min   int    `json:"display_min"`
	Units string `json:"display_units_long"`
}

type LibratoPayload struct {
	Name   string             `json:"name"`
	Time   int64              `json:"measure_time"`
	Val    string             `json:"value,omitempty"`
	Source string             `json:"source,omitempty"`
	User   string             `json:",omitempty"`
	Pass   string             `json:",omitempty"`
	Attr   *LibratoAttributes `json:"attributes,omitempty"`
	Count  *int               `json:"count,omitempty"`
	Sum    *float64           `json:"sum,omitempty"`
	Min    *float64           `json:"min,omitempty"`
	Max    *float64           `json:"max,omitempty"`
}

type LibratoRequest struct {
	Gauges []*LibratoPayload `json:"gauges"`
}

type LibratoOutlet struct {
	// The inbox is used to hold empty buckets that are
	// waiting to be processed. We buffer the chanel so
	// as not to slow down the fetch routine.
	Inbox chan *bucket.Bucket
	// The converter will take items from the inbox,
	// fill in the bucket with the vals, then convert the
	// bucket into a librato metric.
	Conversions chan *LibratoPayload
	// The converter will place the librato metrics into
	// the outbox for HTTP submission. We rely on the batch
	// routine to make sure that the collections of librato metrics
	// in the outbox are homogeneous with respect to their token.
	// This ensures that we route the metrics to the correct librato account.
	Outbox chan []*LibratoPayload
	// How many outlet routines should be running.
	NumOutlets int
	// How many accept routines should be running.
	NumConverters int
	Reader        Reader
	Retries       int
	User          string
	Pass          string
}

func NewLibratoOutlet(mi, mc, mo int) *LibratoOutlet {
	l := new(LibratoOutlet)
	l.Inbox = make(chan *bucket.Bucket, mi)
	l.Conversions = make(chan *LibratoPayload, mc)
	l.Outbox = make(chan []*LibratoPayload, mo)
	return l
}

func (l *LibratoOutlet) Start() {
	go l.Reader.Start(l.Inbox)
	for i := 0; i < l.NumConverters; i++ {
		go l.convert()
	}
	go l.batch()
	for i := 0; i < l.NumOutlets; i++ {
		go l.outlet()
	}
	go l.report()
}

func (l *LibratoOutlet) report() {
	for _ = range time.Tick(time.Second * 2) {
		utils.MeasureI("librato-outlet.inbox", "buckets", int64(len(l.Inbox)))
		utils.MeasureI("librato-outlet.conversions", "payloads", int64(len(l.Conversions)))
		utils.MeasureI("librato-outlet.outbox", "requests", int64(len(l.Outbox)))
	}
}

func (l *LibratoOutlet) convert() {
	for bucket := range l.Inbox {
		if len(bucket.Vals) == 0 {
			fmt.Printf("at=bucket-no-vals bucket=%s\n", bucket.Id.Name)
			continue
		}
		//TODO(ryandotsmith): This is getting out of control.
		//We need a succinct way to building payloads.
		countAttr := &LibratoAttributes{Min: 0, Units: "count"}
		attrs := &LibratoAttributes{Min: 0, Units: bucket.Id.Units}
		switch bucket.Id.User {
		case "librato-production@heroku.com", "dod@heroku.com":
			{
				sum := bucket.Sum()
				count := bucket.Count()
				min := bucket.Min()
				max := bucket.Max()
				l.Conversions <- &LibratoPayload{
					Attr:   attrs,
					User:   bucket.Id.User,
					Pass:   bucket.Id.Pass,
					Time:   ft(bucket.Id.Time),
					Source: bucket.Id.Source,
					Name:   bucket.Id.Name,
					Sum:    &sum,
					Count:  &count,
					Min:    &min,
					Max:    &max,
				}
			}
		default:
			{
				l.Conversions <- &LibratoPayload{Attr: countAttr, User: bucket.Id.User, Pass: bucket.Id.Pass, Time: ft(bucket.Id.Time), Source: bucket.Id.Source, Name: bucket.Id.Name + ".count", Val: fi(bucket.Count())}
				l.Conversions <- &LibratoPayload{Attr: attrs, User: bucket.Id.User, Pass: bucket.Id.Pass, Time: ft(bucket.Id.Time), Source: bucket.Id.Source, Name: bucket.Id.Name + ".sum", Val: ff(bucket.Sum())}
				l.Conversions <- &LibratoPayload{Attr: attrs, User: bucket.Id.User, Pass: bucket.Id.Pass, Time: ft(bucket.Id.Time), Source: bucket.Id.Source, Name: bucket.Id.Name + ".min", Val: ff(bucket.Min())}
				l.Conversions <- &LibratoPayload{Attr: attrs, User: bucket.Id.User, Pass: bucket.Id.Pass, Time: ft(bucket.Id.Time), Source: bucket.Id.Source, Name: bucket.Id.Name + ".max", Val: ff(bucket.Max())}
				l.Conversions <- &LibratoPayload{Attr: attrs, User: bucket.Id.User, Pass: bucket.Id.Pass, Time: ft(bucket.Id.Time), Source: bucket.Id.Source, Name: bucket.Id.Name + ".mean", Val: ff(bucket.Mean())}
			}
		}
		l.Conversions <- &LibratoPayload{Attr: attrs, User: bucket.Id.User, Pass: bucket.Id.Pass, Time: ft(bucket.Id.Time), Source: bucket.Id.Source, Name: bucket.Id.Name + ".median", Val: ff(bucket.Median())}
		l.Conversions <- &LibratoPayload{Attr: attrs, User: bucket.Id.User, Pass: bucket.Id.Pass, Time: ft(bucket.Id.Time), Source: bucket.Id.Source, Name: bucket.Id.Name + ".perc95", Val: ff(bucket.P95())}
		l.Conversions <- &LibratoPayload{Attr: attrs, User: bucket.Id.User, Pass: bucket.Id.Pass, Time: ft(bucket.Id.Time), Source: bucket.Id.Source, Name: bucket.Id.Name + ".perc99", Val: ff(bucket.P99())}
		l.Conversions <- &LibratoPayload{Attr: attrs, User: bucket.Id.User, Pass: bucket.Id.Pass, Time: ft(bucket.Id.Time), Source: bucket.Id.Source, Name: bucket.Id.Name + ".last", Val: ff(bucket.Last())}
		fmt.Printf("measure.bucket.conversion.delay=%d name=%q user=%q time=%d\n", bucket.Id.Delay(time.Now()), bucket.Id.Name, bucket.Id.User, ft(bucket.Id.Time))
	}
}

func (l *LibratoOutlet) batch() {
	ticker := time.Tick(time.Millisecond * 200)
	batchMap := make(map[string][]*LibratoPayload)
	for {
		select {
		case <-ticker:
			for k, v := range batchMap {
				if len(v) > 0 {
					l.Outbox <- v
				}
				delete(batchMap, k)
			}
		case payload := <-l.Conversions:
			index := payload.User + ":" + payload.Pass
			_, present := batchMap[index]
			if !present {
				batchMap[index] = make([]*LibratoPayload, 1, 300)
				batchMap[index][0] = payload
			} else {
				batchMap[index] = append(batchMap[index], payload)
			}
			if len(batchMap[index]) == cap(batchMap[index]) {
				l.Outbox <- batchMap[index]
				delete(batchMap, index)
			}
		}
	}
}

func (l *LibratoOutlet) outlet() {
	for payloads := range l.Outbox {
		if len(payloads) < 1 {
			fmt.Printf("at=%q\n", "empty-metrics-error")
			continue
		}
		//Since a playload contains all metrics for
		//a unique librato user/pass, we can extract the user/pass
		//from any one of the payloads.
		sample := payloads[0]
		reqBody := new(LibratoRequest)
		reqBody.Gauges = payloads
		j, err := json.Marshal(reqBody)
		if err != nil {
			fmt.Printf("at=json-error error=%s user=%s\n", err, sample.User)
			continue
		}
		l.postWithRetry(sample.User, sample.Pass, bytes.NewBuffer(j))
	}
}

func (l *LibratoOutlet) postWithRetry(u, p string, body *bytes.Buffer) error {
	for i := 0; i <= l.Retries; i++ {
		if err := l.post(u, p, body); err != nil {
			fmt.Printf("measure.librato.error user=%s msg=%s attempt=%d\n", u, err, i)
			if i == l.Retries {
				return err
			}
			continue
		}
		return nil
	}
	//Should not be possible.
	return errors.New("Unable to post.")
}

func (l *LibratoOutlet) post(u, p string, body *bytes.Buffer) error {
	defer utils.MeasureT("librato-post", time.Now())
	req, err := http.NewRequest("POST", libratoUrl, body)
	if err != nil {
		return err
	}
	req.Header.Add("Content-Type", "application/json")
	req.Header.Add("User-Agent", "l2met/0")
	req.SetBasicAuth(u, p)
	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	if resp.StatusCode/100 != 2 {
		var m string
		s, err := ioutil.ReadAll(resp.Body)
		if err != nil {
			m = fmt.Sprintf("error=failed-request code=%d", resp.StatusCode)
		} else {
			m = fmt.Sprintf("error=failed-request code=%d resp=body=%s req-body=%s",
				resp.StatusCode, s, body)
		}
		return errors.New(m)
	}
	return nil
}

func ff(x float64) string {
	return strconv.FormatFloat(x, 'f', 5, 64)
}

func fi(x int) string {
	return strconv.FormatInt(int64(x), 10)
}

func ft(t time.Time) int64 {
	return t.Unix()
}
